SSA/Projects/FullyHomomorphicEncryption/PrettySyntax.lean

```diff
@@ -0,0 +1,53 @@
+import SSA.Core.MLIRSyntax.PrettyEDSL
+import SSA.Projects.FullyHomomorphicEncryption.Syntax
+
+import Mathlib.Init.Data.Nat.Lemmas
+
+namespace MLIR.EDSL.Pretty
+open Lean
+
+syntax "poly.add" : MLIR.Pretty.uniform_op
+syntax "return" : MLIR.Pretty.uniform_op
+
+syntax mlir_op_operand " = " "arith.const" "$" noWs "{" term "}" " : " mlir_type : mlir_op
+syntax mlir_op_operand " = " "arith.const" neg_num " : " mlir_type : mlir_op
+macro_rules
+  | `(mlir_op| $v:mlir_op_operand = arith.const $x:neg_num : $t) =>
+      `(mlir_op| $v:mlir_op_operand = "arith.const" () {value = $x:neg_num } : () -> ($t))
+  | `(mlir_op| $v:mlir_op_operand = arith.const ${ $x:term } : $t) => do
+      let ctor := mkIdent ``MLIR.AST.AttrValue.int
+      let x ← `($ctor $x [mlir_type| i64])
+      --                         ^^^^^^^^^^^^^^^^ Is this right? It seems we just ignore the
+      --                                          type annotation
+      `(mlir_op| $v:mlir_op_operand = "arith.const" () {value = $$($x) } : () -> ($t))
+
+syntax mlir_op_operand " = " "poly.monomial" mlir_op_operand,*
+  " : " "("mlir_type,*")" " -> " mlir_type : mlir_op
+macro_rules
+  | `(mlir_op| $v:mlir_op_operand = poly.monomial $xs,* : ($ts,*) -> $t) =>
+      `(mlir_op| $v:mlir_op_operand = "poly.monomial" ($xs,*) : ($ts,*) -> $t)
+
+
+section Test
+variable {q n} [h : Fact (q > 1)]
+
+private def fhe_test_one_lhs := [fhe_com q, n, h | {
+  ^bb0(%a : ! R):
+  --        ^^^ TODO: should `: R` also work here?
```

Comment:
Not sure I understand where the `!` leaks in from!

Reply:
the issue with that was something to do with the mlir type syntax:
```lean
syntax "!" str : mlir_type
syntax "!" ident : mlir_type
syntax ident: mlir_type
syntax "_" : mlir_type

macro_rules
  | `([mlir_type| $x:ident ]) => do
        let (.ident _ xstr _ _) := x.raw | Macro.throwUnsupported
        -- ^^ We use `rawVal` rather than `val`, so that we're not affected by hygiene
        if xstr == "index"
        then
          `(MLIRType.index)
        else if xstr.front == 'i' || xstr.front == 'f'
        then do
          let xstr' := (xstr.drop 1).toString
          match xstr'.toInt? with
          | some _ =>
            let lit := Lean.Syntax.mkNumLit xstr'
            if xstr.front == 'i'
            then `(MLIRType.int .Signless $lit)
            else `(MLIRType.float $lit)
          | none =>
              Macro.throwErrorAt x $ "cannot convert suffix of i/f to int: " ++ xstr.toString
        else Macro.throwErrorAt x $ "expected i<int> or f<int>, found: " ++ xstr.toString
  | `([mlir_type| ! $x:str ]) => `(MLIRType.undefined $x)
  | `([mlir_type| ! $x:ident ]) => `(MLIRType.undefined $(Lean.quote x.getId.toString))
  -- Hardcoded meta-variable
  | `([mlir_type| _ ]) => `(MLIRType.int (φ := 1) Signedness.Signless (Width.mvar ⟨0, by simp⟩))
```

I used the `!` because that's the only way to get a generic type name in here (for dialect-specific types). This seems to match the MLIR documentation: https://mlir.llvm.org/docs/LangRef/#dialect-types